rfd/00XX-agent-census.md

```diff
@@ -0,0 +1,271 @@
+
+In the end, `UpstreamInventoryAgentMetadata.ContainerOrchestrator` will be set to:
+- `kubernetes-eks` if on EKS
+- `kubernetes-gke` if on GKE
+- `kubernetes-unknown` otherwise (AKS, other cloud provider, or no cloud provider)
+
+##### 10. Cloud environment
+
+The only way to determine this seems to be by hitting certain HTTP endpoints specific to each cloud environment:
+- AWS ([docs](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html)): http://169.254.169.254/latest/
+- GCP ([docs](https://cloud.google.com/compute/docs/metadata/overview#parts-of-a-request)): http://metadata.google.internal/computeMetadata/v1/
+- Azure ([docs](https://learn.microsoft.com/en-us/azure/virtual-machines/linux/instance-metadata-service?tabs=linux#access-azure-instance-metadata-service)): http://169.254.169.254/metadata/instance?api-version=2021-02-01
+
+`UpstreamInventoryAgentMetadata.CloudEnvironment` will be set to:
+- `aws` if on AWS
+- `gcp` if on GCP
+- `azure` if on Azure
+
+#### Development plan
+
+The above work will be divided in the following tasks:
+
+1. Add new message type `AgentMetadataEvent` to PreHog.
+2. Add new message type `UpstreamInventoryAgentMetadata`. These will be empty initially, except for the fields common with `UpstreamInventoryHello`. This is okay since being empty means that the field could not be determined, which may happen anyways.
+3. Extend auth server to convert `UpstreamInventoryAgentMetadata` messages to `AgentMetadataEvent` messages and push them to PreHog.
+4. Gradually instrument & add new code that fills each new `UpstreamInventoryAgentMetadata` message field.
+
+We could decide to do step 4 together with step 2 if we don't want to risk adding fields to `UpstreamInventoryAgentMetadata` that possibly won't be used in the end (if for some reason we figure out they can/should not be tracked).
+However, step 4 requires several changes (Go code & files used by the multiple installation methods) which we may want to review separately.
+A similar reasoning also applies to step 1 since each field in `AgentMetadataEvent` should only exist if it also exists in `UpstreamInventoryAgentMetadata`.
+
+### Security
+
+Detecting the __9. Container orchestrator__ and __10. Cloud environment__ requires hitting certain HTTP endpoints.
+This may be considered too intrusive, so we have to make a decision on whether we really want to track it and argue why it's okay to do so.
+
+### UX
+
+Data analysis and visualization are not a goal for this RFD, so no UX concerns for now.
+
+### Open questions
+
+- Which alternative should we use for __5. Host architecture__?
+- Should we consider [Teleport AMIs](https://github.com/gravitational/teleport/tree/6f9ad9553a5b5946f57cb35411c598754d3f926b/examples/aws/terraform/AMIS.md) an installation method?
```

Comment:
I think this would be valuable. There is a chunk of hardening and maintenace that goes into our AMIs, and it'd be nice to know at what level we should continue to invest in these.

I don't feel strongly that this should be included in the initial scope.

Reply:
I moved them to the list of install methods we may want to track once the initial phase of this project has been implemented: b39a4a5897fd849222005bd09252193a410c1cfc